dbms/src/DataTypes/DataTypeDateTime64.cpp

```diff
@@ -0,0 +1,257 @@
+#include <DataTypes/DataTypeDateTime64.h>
+
+#include <Columns/ColumnDecimal.h>
+#include <Columns/ColumnVector.h>
+#include <Columns/ColumnsNumber.h>
+#include <Common/assert_cast.h>
+#include <Common/typeid_cast.h>
+#include <common/DateLUT.h>
+#include <DataTypes/DataTypeFactory.h>
+#include <Formats/FormatSettings.h>
+#include <Formats/ProtobufReader.h>
+#include <Formats/ProtobufWriter.h>
+#include <IO/Operators.h>
+#include <IO/ReadHelpers.h>
+#include <IO/WriteBufferFromString.h>
+#include <IO/WriteHelpers.h>
+#include <IO/parseDateTimeBestEffort.h>
+#include <Parsers/ASTLiteral.h>
+
+#include <optional>
+#include <string>
+
+namespace
+{
+using namespace DB;
+
+static inline void readText(DateTime64 & x, UInt32 scale, ReadBuffer & istr, const FormatSettings & settings, const DateLUTImpl & time_zone, const DateLUTImpl & utc_time_zone)
+{
+    switch (settings.date_time_input_format)
+    {
+        case FormatSettings::DateTimeInputFormat::Basic:
+            readDateTime64Text(x, scale, istr, time_zone);
+            return;
+        case FormatSettings::DateTimeInputFormat::BestEffort:
+            parseDateTime64BestEffort(x, scale, istr, time_zone, utc_time_zone);
+            return;
+    }
+}
+}
+
+namespace DB
+{
+
+DataTypeDateTime64::DataTypeDateTime64(UInt32 scale_, const std::string & time_zone_name)
+    : DataTypeDecimalBase<DateTime64>(maxDecimalPrecision<DateTime64>() - scale_, scale_),
```

Comment:
I've asked @4ertus2 and think this is not correct. We should use `maxDecimalPrecision<DateTime64>()` for precision, not `maxDecimalPrecision<DateTime64>() - scale_`. Cause precision is a number of digits in total, not before dot.

Reply:
This is still the case.